### ISSUE TITLE
HTTPMatchRequest.uri and .headers are not treated as AND

### ISSUE THREAD

**Issue by esnible** [istio/istio#5518]
Given a _VirtualService_ with

```
  http:
  - match:
    - uri:
        prefix: /hello
    - headers:
        user-agent:
          prefix: Mozilla
    route:
    - destination:
        host: helloworld-service
        port:
          number: 8080
```

The generated output (as seen by curl istio-pilot.istio-system:15007/debug/adsz ) is 

```
                      "routes": [
                        {
                          "decorator": {
                            "operation": "guestbook"
                          },
                          "match": {
                            "prefix": "/hello"
                          },
                          "route": {
                            "cluster": "outbound|http||helloworld-service.default.svc.cluster.local",
                            "timeout": "0.000s",
                            "use_websocket": false
                          }
                        },
                        {
                          "decorator": {
                            "operation": "guestbook"
                          },
                          "match": {
                            "headers": [
                              {
                                "name": "user-agent",
                                "regex": true,
                                "value": "^Mozilla.*"
                              }
                            ],
                            "prefix": "/"
                          },
                          "route": {
                            "cluster": "outbound|http||helloworld-service.default.svc.cluster.local",
                            "timeout": "0.000s",
                            "use_websocket": false
                          }
                        }
                      ]
```

That's incorrect: there should be one generated Envoy route that matches on /hello and headers.

If I visit the service without the special header I get a valid response:

```
$ curl -H "User-Agent: curl" http://169.61.14.73:31380/hello/abc
{"hostname":"helloworld-service-v1-5dbf6d5649-dc5bw","greeting":"Hello abc from helloworld-service-v1-5dbf6d5649-dc5bw with 1.0","version":"1.0"}
```

I expected the `match` would not succeeded unless BOTH the headers and URI matched.

**Comment by rshriram** [comment]
Syntax is incorrect. You have two different match conditions.

**Comment by linsun** [comment]
@rshriram that is on purpose.  Could you suggest how to do match /hello path and mobile as user-gent?  cc @frankbu

**Comment by esnible** [comment]
@rshriram I'm looking at https://preliminary.istio.io/docs/reference/config/istio.networking.v1alpha3.html#HTTPMatchRequest which says "a set of criterion to be met in order for the rule to be applied to the HTTP request. "  I believe that all criteria in an HTTPMatchRequest must be met and I've put both a `uri` and `headers` under the same `match`.  Can you provide or point me to an example with correct syntax that matches when the URI begins with /hello and the headers match?

**Comment by rshriram** [comment]
You have two different match conditions (OR semantics) one with uri and another with header.

**Comment by frankbu** [comment]
Ed, it should be like this:
```
  - match:
    - uri:
        prefix: /hello
      headers:
        user-agent:
          prefix: Mozilla
```

**Comment by linsun** [comment]
Ok, this does work.  Probably would be valuable to have these 2 examples documented.

**Comment by denniseffing** [comment]
Are there any plans to include this example in the official docs?

**Comment by zhengweisk** [comment]
不错。记录下

uri and header 可以一起 为真时，才转发到相应的后端。

**Comment by lisenet** [comment]
I had the same problem, and getting rid of the hyphen fixed it for me.

It would make sense to update Istio documentation. There is this line in the "headers" sections (see https://istio.io/v1.8/docs/reference/config/networking/virtual-service/#HTTPMatchRequest or the latest 1.9 docs):

"Note: The keys uri, scheme, method, and authority will be ignored."

I tried it, it did not work, I therefore assumed that it was not supported. I removed the hyphen as per suggestion above and it started to work.
